File: 0020-valid-parentheses/0020-valid-parentheses.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s)<=1:
            return False
        stack = deque()
        dict = {"]":"[", "}":"{", ")":"("}
        for char in s:
            if char in dict.values():
                stack.append(char)
            elif char in dict.keys(): 
                if not stack or dict[char]!=stack.pop():
                    return False
            else:
                return False
        return len(stack)==0

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    
    private HashMap<Character, Character> mappings;
    
    public Solution(){
        this.mappings = new HashMap<Character, Character>();
        this.mappings.put(')', '(');
        this.mappings.put(']', '[');
        this.mappings.put('}', '{');
    }
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i=0; i<s.length(); i++){
            char curr = s.charAt(i);
            if(this.mappings.containsKey(curr)){
                char topElement = stack.empty() ? '#' : stack.pop();
                if(topElement!=this.mappings.get(curr)){
                    return false;
                }
            } else{
                stack.push(curr);
            }
        }
        return stack.isEmpty();
    }
}